libraries/AP_HAL_ChibiOS/Scheduler.cpp

```diff
@@ -440,7 +441,14 @@ void Scheduler::_monitor_thread(void *arg)
         }
         if (loop_delay >= 500 && !sched->in_expected_delay()) {
             // at 500ms we declare an internal error
-            INTERNAL_ERROR(AP_InternalError::error_t::main_loop_stuck);
+            AP::internalerror().error(AP_InternalError::error_t::main_loop_stuck, hal.util->persistent_data.semaphore_line);
+            /*
+              if we are armed and get this condition then it is likely
```

Comment:
I disagree with the "likely" here. I have seen main loop stuck for other reasons, some of which are recoverable. For instance a different kind of bug where someone hasn't done a proper EXPECTED_DELAY(). At the point you nuke the mutex we really want to be completely out of options, because once you do it the firmware will be in an unknown state which you cannot recover from. Could you maybe instead try this the first time you were about to watchdog (and reset the watchdog timer) and then second time if everything is still toast, watchdog at that point?

Reply:
we're already 0.5s of no main loop update at this point. If we leave it much longer we start to lose our attitude estimate.
Leaving till we're just about to watchdog would be over 1.8s, by which time many VTOL vehicles are already completely hosed

Reply:
we could make the threshold higher when disarmed, but I think 500ms is already a long time to wait while armed
EXPECTED_DELAY() only happens while disarmed

Reply:
higher threshold when disarmed makes sense to me, even better to only do it when flying - there is usually a log pause when logging gets going which can lead to some longer than average delays